### DISCUSSION TITLE
Detox Test Flakiness

### DISCUSSION THREAD

**Discussion by jchan29** [wix/Detox#3460] [Q&A]
Hi all, our team is experiencing flakiness when running our detox tests (which only includes a test that verifies a login user flow) locally. The detox server unexpectedly loses connection with our Android emulator and sometimes is unable to connect to our test app. At times, we also get errors about insufficient storage. We're open to any suggestions and resources you can share.

On consecutive runs, the errors are inconsistent and include:

    The app has unexpectedly disconnected from Detox server.

      1 | /* eslint-disable no-undef*/
      2 | beforeAll(async () => {
    > 3 |   await device.launchApp();
        |                ^
      4 |   await device.setURLBlacklist(['.*pndsn.com/v2/subscribe.*']);
      5 | });
      6 | 
 
    The app has unexpectedly disconnected from Detox server.

      15 |       .toBeVisible()
      16 |       .withTimeout(500);
    > 17 |     await element(by.text('Provider')).tap();

```
detox[64196] INFO:  [APP_STATUS] App synchronization debug: The app appears to be idle!
```
```
detox[64196] INFO:  [APP_STATUS] App synchronization debug: 
The app is busy, due to: 
	 - Enqueued timers
detox[64196] WARN:  [PENDING_REQUESTS] The app has not responded to the network requests below:
  (id = 2) invoke: {"target":{"type":"Class","value":"com.wix.detox.espresso.EspressoDetox"},"method":"perform","args":[{"type":"Invocation","value":{"target":{"type":"Class","value":"androidx.test.espresso.Espresso"},"method":"onView","args":[{"type":"Invocation","value":{"target":{"type":"Class","value":"com.wix.detox.espresso.DetoxMatcher"},"method":"matcherForContentDescription","args":["login_info_text"]}}]}},{"type":"Invocation","value":{"target":{"type":"Class","value":"com.wix.detox.espresso.DetoxViewActions"},"method":"click","args":[]}}]}

That might be the reason why the test "Auth test should login to the app" has timed out.
```
```
detox[59663] ERROR: [EXEC_FAIL, #45] Failure [INSTALL_FAILED_INSUFFICIENT_STORAGE]
```
```
    DetoxRuntimeError: Detox can't seem to connect to the test app(s)!

    HINT: 

    Have you forgotten to call 'device.launchApp()' in the beginning of your test?
    Refer to our troubleshooting guide, for full details: https://github.com/wix/Detox/blob/master/docs/Troubleshooting.RunningTests.md#tests-execution-hangs

    ---
    The following package could not be delivered:

    {
      type: 'reactNativeReload',
      params: {},
      messageId: -1000
    }

       6 | 
       7 | beforeEach(async () => {
    >  8 |   await device.reloadReactNative();
         |                ^
       9 | });
      10 | 
```
```
    Test Failed: 'at least 75 percent of the view's area is displayed to the user.' doesn't match the selected view.
    Expected: at least 75 percent of the view's area is displayed to the user.
         Got: null
```

**Comment by portyu9** [comment]
Hi @jchan29 

There are several symptoms in the logs, but one concrete infrastructure failure should be fixed first:

```text
INSTALL_FAILED_INSUFFICIENT_STORAGE
```

Check the emulator storage:

```bash
adb shell df
```

Then wipe/recreate the AVD, increase its storage allocation, or remove stale apps/data.

Once APK installation or reinstallation becomes unreliable, later errors such as Detox disconnects, failed reloads, pending commands, or inability to reconnect can be secondary symptoms.

Two other things are worth changing while debugging:

**1. A 500 ms visibility timeout is extremely short.**

If you're using something like:

```js
withTimeout(500)
```

during login/navigation, use a realistic synchronization timeout instead.

**2. Avoid depending on `reloadReactNative()` while isolating this.**

Current Detox documentation warns that `reloadReactNative()` can misbehave or crash in some situations. Prefer:

```js
await device.launchApp({ newInstance: true })
```

while debugging.

After emulator storage/install health is fixed, if Detox still waits indefinitely, investigate synchronization separately using Detox's synchronization diagnostics.

I would not treat all of the pasted errors as one networking issue.

Please mark this answer as accepted if it helped, thank you.
